python/cudf/cudf/_lib/pylibcudf/scalar.pyx

```diff
@@ -0,0 +1,99 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.
+
+cimport pyarrow.lib
+from cython cimport no_gc_clear
+from cython.operator cimport dereference
+from libcpp.utility cimport move
+
+import pyarrow.lib
+
+from rmm._lib.memory_resource cimport get_current_device_resource
+
+from cudf._lib.cpp cimport aggregation
+from cudf._lib.cpp.column.column_view cimport column_view
+from cudf._lib.cpp.reduce cimport cpp_reduce
+from cudf._lib.cpp.scalar.scalar cimport scalar
+from cudf._lib.cpp.types cimport data_type
+
+from .column cimport Column
+from .interop cimport from_arrow
+from .table cimport Table
+from .types cimport DataType
+
+
+# The DeviceMemoryResource attribute could be released prematurely
+# by the gc if the DeviceScalar is in a reference cycle. Removing
+# the tp_clear function with the no_gc_clear decoration prevents that.
+# See https://github.com/rapidsai/rmm/pull/931 for details.
+@no_gc_clear
+cdef class Scalar:
+    """A scalar value in device memory."""
+    # Unlike for columns, libcudf does not support scalar views. All APIs that
+    # accept scalar values accept references to the owning object rather than a
+    # special view type. As a result, pylibcudf.Scalar has a simpler structure
+    # than pylibcudf.Column because it can be a true wrapper around a libcudf
+    # column
+
+    def __init__(self, pyarrow.lib.Scalar value=None):
+        self.mr = get_current_device_resource()
+
+        if value is None:
+            # TODO: This early return is not something we really want to
+            # support, but it here for now to ease the transition of
+            # DeviceScalar.
+            return
+
+        # Convert the value to a cudf object via pyarrow
+        arr = pyarrow.lib.array([value.as_py()], type=value.type)
```

Comment:
More effort, but I think we can use the C++-level [`ArrayBuilder`](https://arrow.apache.org/docs/cpp/api/builder.html#_CPPv4N5arrow12ArrayBuilder12AppendScalarERK6Scalar) interface to get an arrow array from which we can make a table to pass to libcudf. If we push it all onto C++ then we can probably avoid some back and forth too

Reply:
This (untested) might be a useful starting point:

```c++
std::unique_ptr<scalar> from_arrow(arrow::Scalar const &scalar, rmm::mr::device_memory_resource* mr) {
  auto b = arrow::MakeBuilder(scalar.type);
  if (b.ok()) {
    auto builder = std::move(b.ValueOrDie());
    arrow::Status status = builder->AppendScalar(scalar);
    if (status != arrow::Status::OK()) {
      CUDF_FAIL("Arrow ArrayBuilder::AppendScalar failed");
    }
    auto array_result = builder->Finish();
    if (array_result.ok()) {
      auto array = array_result.ValueOrDie();
      auto type = detail::arrow_to_cudf_type(*array->type().get());
      auto col = detail::get_column(*array.get(), type, true, cudf::get_default_stream(), mr);
      auto cv = col->view();
      return detail::get_element(cv, 0, cudf::get_default_stream(), mr);
    } else {
      CUDF_FAIL("Arrow ArrayBuilder::Finish failed");
    }
  } else {
    CUDF_FAIL("Arrow MakeBuilder failed");
  }
}
```

Reply:
I was hoping to avoid the C++ API, but at this point I think you're right that the most straightforward solution is to implement at this level. I'll look into it.